microsoft/testsuites/device_passthrough/functional_tests.py

```diff
@@ -0,0 +1,89 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+from pathlib import Path
+from typing import Any, Dict, cast
+
+from lisa import (
+    Environment,
+    Logger,
+    Node,
+    TestCaseMetadata,
+    TestSuite,
+    TestSuiteMetadata,
+)
+from lisa.sut_orchestrator import CLOUD_HYPERVISOR
+from lisa.sut_orchestrator.libvirt.ch_platform import CloudHypervisorPlatform
+from lisa.sut_orchestrator.libvirt.context import get_node_context as get_ch_context
+from lisa.sut_orchestrator.libvirt.schema import BaseLibvirtNodeSchema
+from lisa.testsuite import TestResult, simple_requirement
+from lisa.tools import Lspci
+from lisa.util import LisaException, SkippedException
+
+
+@TestSuiteMetadata(
+    area="device_passthrough",
+    category="functional",
+    description="""
+    This test suite is for testing device passthrough functional tests.
+    """,
+)
+class DevicePassthroughFunctionalTests(TestSuite):
+    @TestCaseMetadata(
+        description="""
+            Check if passthrough device is visible to guest.
+            This testcase support only on CLOUD_HYPERVISOR
+            platform of LISA.
+        """,
+        priority=4,
+        requirement=simple_requirement(
+            supported_platform_type=[CLOUD_HYPERVISOR],
+        ),
+    )
+    def verify_device_passthrough_on_guest(
+        self,
+        log: Logger,
+        node: Node,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        variables: Dict[str, Any],
+    ) -> None:
+        ctx = get_ch_context(node)
```

Comment:
The context is not used after assertion.

Reply:
Yes, context is to check if any device-passthrough is set or not. I can check that with extended runbook too. Will remove context now

Reply:
It may hide some bugs. The case should check if the platform defined devices or not, instead of the actual devices exists or not.

Reply:
@squirrelsc 

Flow is like, based on runbook, platform will set the node context under `passthrough_devices` and if that is set, we want to verify that what are the devices assigned to libvirt domain.

Reason on verifying actual devices is present in the VM is

- Make sure devices are visible to guest. If there is a case where libvirt domain may come up but device is not correctly passthrough
- Run functional aspect based on device type (NIC, GPU etc), This is considered for upcoming work. We can extend the same testcase to verify this. Current scope is limited for the testcase, but we will extend this in upcoming work cycle.
- bdf for pci device under Guest VM would be different. So only way to verify is from runbook. Check the node requirement to get the device count and device type (pool type) and get what is the vendor/device id for given device type so We can check on VM that whatever runbook has defined as devices are visible to guest or not.

I am still a bit unclear when you said it may hide some bugs. Could you please explain.

Reply:
For most test cases, they verify if the "actual" value is the same as the "expected" value. In this case, node context has the actual value, and runbook has the expected value.
It should be skipped only the "expected" value doesn't meet the test criteria. It shouldn't skip the case, if the "actual" value by the "actual" value. Because if the expected value expects 1, but actual value is 0. It should be failed, instead of skipped.

Reply:
@squirrelsc 

Sure CHi, I will remove node context check. We can do the same check on extended runbook. 

If "Expected" device from runbook is not present, we are failing the test anyway.